### DISCUSSION TITLE
Best practice to overriding default value from a definition in another definition?

### DISCUSSION THREAD

**Discussion by leitzler** [cuelang/cue#989] [Q&A] [answered]
Using `v0.4.0-rc.1`:

I'm trying to provide a "new default value" for a nested definition and the error message is slightly confusing. Is there a more natural way to achieve this?

E.g.: I want a new definition `#Foo` that has a field `bar` that must be a `#AB`, but IF it is a `#B` it should have a different default value.

```cue
package foobar

#AB: {
	"a"
} | {
	b: #B
}

#B: {
	bf: string | *"default"
}

#Foo: {
	// This works:
	//bar: {b: #B & {bf: "newDefault"}}
	//  $ cue eval foo.cue --out yaml
	//  bar:
	//    b:
	//      bf: newDefault

	// This also works:
	//bar: #AB
	//
	//   $ cue eval foo.cue --out yaml
	//   bar:
	//     b:
	//       bf: default
	//

	// Doesn't work:
	bar: #AB | {
		b: #B & {bf: "newDefault"}
	}
	//   $ cue vet -c foo.cue
	//   bar: incomplete value {b:{bf:*"default" | string}} | {b:{bf:"newDefault"}}
	//
}

// I want to enforce a default value for "bf" via "& #Foo"
{
	bar: b: #B
} & #Foo
```

**Comment by mpvl** [comment, marked as answer]
In general, a good approach is to not have any defaults in the base template, and then add them later as one desires.

In this case, the issue is that `bar` introduces a disjunction without a default. To ensure it "reaches" the `newDefault` value, you could write
```
	bar: #AB | *{ // <---- NOTE THE ADDED DEFAULT MARKER 
		b: #B & {bf: "newDefault"}
	}
```

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/989.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
